onadata/apps/api/viewsets/open_id_connect_viewset.py

```diff
@@ -0,0 +1,213 @@
+import secrets
+
+from django.conf import settings
+from django.contrib.auth.models import User
+from django.core.cache import cache
+from django.http import (
+    HttpResponseBadRequest, HttpResponseRedirect)
+from django.utils.translation import ugettext as _
+
+import jwt
+from rest_framework import viewsets
+from rest_framework.decorators import action
+from rest_framework.permissions import AllowAny
+from rest_framework.renderers import TemplateHTMLRenderer
+from rest_framework.response import Response
+
+from onadata.apps.main.models import UserProfile
+from onadata.libs.utils.open_id_connect_tools import OpenIDHandler
+
+
+class OpenIDConnectViewSet(viewsets.ViewSet):
+    """
+    OpenIDConnectViewSet: Handles OpenID connect authentication
+
+    This OpenID Connect authentication flow requires the OpenID
+    provider to provide the optional 'given_name', 'family_name' and 'email'
+    claims in their payload
+    """
+    permission_classes = [AllowAny]
+    authentication_classes = []
+    renderer_classes = (TemplateHTMLRenderer, )
+
+    @action(methods=['GET'], detail=False)
+    def initiate_oidc_flow(  # pylint: disable=no-self-use
+            self, request, **kwargs):
+        """
+        This endpoint initiates the OpenID Connect Flow by generating a request
+        to the OpenID Connect Provider with a cached none for verification of
+        the returned request
+        """
+        provider_config, open_id_provider = retrieve_provider_config(
+            **kwargs)
+
+        if provider_config:
+            nonce = secrets.randbits(16)
+            scope = 'openid'
+            response_type = 'code'
+            response_mode = ''
+
+            # Cache the nonce for verification on callback
+
+            if open_id_provider == 'msft':
+                scope = 'openid profile'
+                response_type = 'id_token'
+                response_mode = 'form_post'
+                cache.set(nonce, 'msft')
+            elif open_id_provider == 'wfp':
+                cache.set(nonce, 'wfp')
+
```

Comment:
Can all this come from a configuration?

```python
# local_settings.py

OPEN_ID_PROVIDER = {
   'msft': { 
        'scope': 'openid profile',
        'nonce': 'msft',
        'response_type': 'id_token',
        'response_mode': 'form_post',
    },
}
```
Then you access it as:

```python
from django.conf import settings

...

provider = getattr(settings, 'OPEN_ID_PROVIDER', {}).get('open_id_connect_provider', {})

scope = provider.get('scope', 'scope')
response_type = provider.get('response_type', 'code')
...

```

Reply:
Yes, i'll make this change

Reply:
Changes made in latest commit